toolchain/parse/typed_nodes.h

```diff
@@ -677,6 +677,39 @@ struct IfExprElse {
   AnyExprId else_result;
 };
 
+// Choice nodes
+// ------------
+
+using ChoiceIntroducer = LeafNode<NodeKind::ChoiceIntroducer>;
+
+struct ChoiceSignature {
+  static constexpr auto Kind =
+      NodeKind::ChoiceDefinitionStart.Define(NodeCategory::None);
+
+  ChoiceIntroducerId introducer;
+  llvm::SmallVector<AnyModifierId> modifiers;
+  AnyNameComponentId name;
+  std::optional<ImplicitParamListId> implicit_params;
+  std::optional<TuplePatternId> params;
+};
+
+using ChoiceDefinitionStart = ChoiceSignature;
+
+using ChoiceAlternativeListComma =
+    LeafNode<NodeKind::ChoiceAlternativeListComma>;
+
+struct ChoiceDefinition {
+  static constexpr auto Kind =
+      NodeKind::ChoiceDefinition.Define(NodeCategory::Decl);
+
+  ChoiceDefinitionStartId signature;
+  struct Alternative {
+    AnyNameComponentId name;
```

Comment:
I think we only want to allow simple identifier names here. I don't have a strong preference as to whether that's validated in the parser or in check, but either way, we should have a test covering this case, for example:
```
choice X { A.B }
```